kernel/src/util/iovec.rs

```diff
@@ -37,92 +39,192 @@ impl TryFrom<UserIoVec> for IoVec {
-    /// 2. `self.is_empty()` is `true`.
-    pub fn write_exact_to_user(&self, src: &[u8]) -> Result<()> {
-        assert_eq!(src.len(), self.len);
-        assert!(!self.is_empty());
+    /// This method returns [`Errno::EFAULT`] if a page fault occurs.
+    fn write(&mut self, reader: &mut VmReader<'_, Infallible>) -> Result<usize>;
+
+    /// Calculates the length of space available to write.
+    fn sum_lens(&self) -> usize;
 
-        CurrentUserSpace::get().write_bytes(self.base, &mut VmReader::from(src))
+    /// Checks if the space available to write is empty.
+    fn is_empty(&self) -> bool {
+        self.sum_lens() == 0
     }
+}
 
-    /// Reads bytes to the `dst` buffer
-    /// from the user space buffer pointed by the `IoVec`.
-    ///
-    /// If successful, returns the length of actually read bytes.
-    pub fn read_from_user(&self, dst: &mut [u8]) -> Result<usize> {
-        let len = self.len.min(dst.len());
-        CurrentUserSpace::get().read_bytes(self.base, &mut VmWriter::from(&mut dst[..len]))?;
-        Ok(len)
+impl<'a> MultiRead for VmReaderArray<'a> {
+    fn read(&mut self, writer: &mut VmWriter<'_, Infallible>) -> Result<usize> {
+        let mut total_len = 0;
+
+        for reader in &mut self.0 {
+            let copied_len = reader.read_fallible(writer)?;
+            if copied_len == 0 {
+                break;
+            }
+            total_len += copied_len
+        }
+        Ok(total_len)
     }
 
-    /// Writes bytes from the `src` buffer
-    /// to the user space buffer pointed by the `IoVec`.
-    ///
-    /// If successful, returns the length of actually written bytes.
-    pub fn write_to_user(&self, src: &[u8]) -> Result<usize> {
-        let len = self.len.min(src.len());
-        CurrentUserSpace::get().write_bytes(self.base, &mut VmReader::from(&src[..len]))?;
-        Ok(len)
+    fn sum_lens(&self) -> usize {
+        self.0.iter().map(|vm_reader| vm_reader.remain()).sum()
     }
 }
 
-/// Copies IO vectors from user space.
-pub fn copy_iovs_from_user(start_addr: Vaddr, count: usize) -> Result<Box<[IoVec]>> {
-    let mut io_vecs = Vec::with_capacity(count);
+impl<'a> MultiRead for VmReader<'a> {
+    fn read(&mut self, writer: &mut VmWriter<'_, Infallible>) -> Result<usize> {
+        Ok(self.read_fallible(writer)?)
+    }
 
-    let user_space = CurrentUserSpace::get();
-    for idx in 0..count {
-        let addr = start_addr + idx * core::mem::size_of::<UserIoVec>();
-        let uiov = user_space.read_val::<UserIoVec>(addr)?;
-        let iov = IoVec::try_from(uiov)?;
-        io_vecs.push(iov);
+    fn sum_lens(&self) -> usize {
+        self.remain()
+    }
+}
+
+impl<'a> MultiWrite for VmWriterArray<'a> {
+    fn write(&mut self, reader: &mut VmReader<'_, Infallible>) -> Result<usize> {
+        let mut total_len = 0;
+
+        for writer in &mut self.0 {
+            let copied_len = writer.write_fallible(reader)?;
+            if copied_len == 0 {
```

Comment:
From https://github.com/asterinas/asterinas/pull/1294#discussion_r1759650046:
> > I think you mean we want to exit the loop when `writer` becomes empty? That sounds right to me.
> 
> I think this is the only reason that `read_fallible` may return `Ok(0)`. When creating `VmReaderArray`, we have already skipped empty readers.

This statement is just wrong.

We may call the `write` method multiple times. For example, in `ring_buffer.rs`:
```rust
            let mut len = writer.write(&mut reader)?;
            len += writer.write(&mut reader)?;
```
In the second call to the `write` method, we may see an empty `writer` and `copied_len == 0`. Obviously, we can't break the loop because the `reader` may not be empty.

Reply:
Thanks for your comment! I have changed the break condition as `if !writer.has_avail()` and if `!reader.has_remain()`.